src/language_features/document_symbol.rs

```diff
@@ -332,25 +332,43 @@ fn next_or_prev_symbol_details<T: Symbol<T> + 'static>(
         let symbol_position =
             get_kakoune_position_with_fallback(&meta.buffile, symbol_position, ctx);
 
-        if exceeds(
-            symbol_position,
-            kind,
-            params.position,
-            params.search_next,
-            &params.symbol_kind,
-        ) {
-            return Some((
-                filename.to_string(),
+        let symbol_name = symbol.name().to_owned();
+
+        // Order of checking the parent symbol for exceeds() matters.
+        // If we're searching forward then check if the parent symbol exceeds() the current
+        // symbol first. If we're searching backwards then check if any of children symbols
+        // exceeds() the current symbol first
```

Comment:
These three sentences are kind of redundant because they only repeat what the code is doing.
What would be much more interesting is *why* there is a difference between forward and backward search.

here is my attempt

```rust
// Assume that children always have a starting position higher than (or equal to) their
// parents.  This means that when searching for the node with the next-higher position
// (anywhere in the tree) we need to check the parent first.  Conversely, when looking
// for the node with the next-lower position, we need to check children first, because
// they have a higher position than their parents.
```

Maybe that's a bit too wordy. It's probably not necessary to spend too much on explaining that.
The key thing is the assumption about parent/child positions, everything else can be derived from it.